HW-2 MVP 패턴 적용

BeokBeok opened the pull request:
google blueprint 오픈 소스 참고하여 MVP 패턴 적용하였습니다.

Commit:
HW-2 함수 내 return 만 있는 경우, 중괄호 제거

BeokBeok committed:
HW-2 MVP 패턴 적용

- Google blueprint 오픈소스 참고
https://github.com/googlesamples/android-architecture/tree/todo-mvp-kotlin/

BeokBeok committed:
HW-2 Market 관련 버튼 추가

BeokBeok committed:
HW-2 Market 관련 버튼 추가에 따른 수치 환산 로직 추가

NackUn commented on BeokBeok/app/src/main/java/com/architecturestudy/upbitmarket/UpBitFragment.kt:
```diff
@@ -0,0 +1,97 @@
+
+class UpBitFragment : BaseFragment(), UpBitContract.View {
+
+    private lateinit var upBitPresenter: UpBitPresenter
```
upBitPresenter 어디에 쓰는 건가요??

BeokBeok replied:
@NackUn 
뷰에서 프레젠터를 통하여 이벤트를 발생 시키고
해당 이벤트에 대한 데이터를 프레젠터에서 받아서 뷰에 전달 시키기 위해 사용합니다.

NackUn replied:
네! 친절한 설명 감사드립니다.

NackUn commented on BeokBeok/app/src/main/java/com/architecturestudy/upbitmarket/UpBitFragment.kt:
```diff
@@ -0,0 +1,97 @@
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        upBitPresenter = UpBitPresenter(UpBitRepository.getInstance(UpBitRetrofitDataSource.getInstance()), this)
```
presenter = UpBitPresenter(UpBitRepository.getInstance(UpBitRetrofitDataSource.getInstance()), this) 이렇게 해주면 안되나요?

BeokBeok replied:
@NackUn 
UpBitRepository 에서 RemoteDataSource (Retrofit) 를 통하여 Retrofit 통신을 하고 난 데이터를 전달받기 위해 사용하였습니다.
단순하게 UpBitRepository에서 Retrofit 통신을 해도 되지만,
추후 Local Data Source 를 사용할 수도 있기 때문에 위와 같은 방식으로 구현하였습니다.

NackUn replied:
그렇군요! 배워갑니다. 제가 아직 미숙해서 조금 더 읽어보고 리뷰 마치도록 하겠습니다.

NackUn commented on BeokBeok/app/src/main/java/com/architecturestudy/data/source/UpBitDataSource.kt:
```diff
@@ -0,0 +1,24 @@
+
+interface UpBitDataSource {
+    interface GetTickerCallback {
+        fun onThickerLoaded(marketPrice: List<UpBitTicker>)
```
onThickerLoaded 함수명 오타이신 것 같아서요

BeokBeok replied:
@NackUn 
함수명 변경하였습니다.
700c9d6

NackUn replied:
확인했습니다. Mentor Needed 했습니다. 고생하셨습니다.

BeokBeok committed:
HW-2 함수명 수정 (onThickerLoaded -> onTickerLoaded)

NackUn commented on BeokBeok/app/src/main/java/com/architecturestudy/data/source/UpBitRetrofitDataSource.kt:
```diff
@@ -0,0 +1,75 @@
+    private fun getTickers(market: List<String?>?, callback: UpBitDataSource.GetTickerCallback) {
+        retrofit.getTicker(market).enqueue(object : Callback<List<UpBitTicker>> {
+            override fun onResponse(call: Call<List<UpBitTicker>>, response: Response<List<UpBitTicker>>) {
+                val responseTicker = response.body()
```
여기에 Elvis operator ( ?: )를 이용해보시는게 어떠실까요??

BeokBeok replied:
@NackUn 
엘비스 연산자는 default 값을 null 이 아닌 다른 값을 지정할 때 사용합니다.
해서, 이 부분에서는 엘비스 연산자의 의도와는 맞지 않다고 생각합니다.

NackUn replied:
아하 그렇군요 그런 의도인지는 몰랐네요 감사합니다!

BeokBeok commented:
@NackUn 
Request changes로 변경 부탁 드립니다.

NackUn reviewed (changes requested):
확인했습니다.

BeokBeok committed:
HW-3 패키지명 변경 (common -> base)

BeokBeok committed:
HW-3 UpBit 이름 변경 (Upbit)

BeokBeok committed:
HW-3 Retrofit 관련 데이터 클래스 정리

- null 초기화 코드 제거

BeokBeok committed:
HW-3 Retrofit 관련 인터페이스 분리 및 BaseUrl 범위 변경

BeokBeok committed:
HW-3 코드 일괄 정리 (1차)

- prefix 관련 enum 처리
- presenter 변수 타입 val로 변경 및 lazy 초기화 적용
- 함수 파라미터 관련 convention 준수
- getInstance 대신 invoke 로 변경
- onDataNotAvailable 인자 변경 (String -> Throwable)
- UpBit -> Upbit 로 추가 변경

BeokBeok committed:
HW-3 AndroidX 마이그레이션 적용 및 코틀린 버전 업데이트

BeokBeok committed:
HW-3 코드 일괄 정리 (2차)

- invoke 함수 보강 (싱글턴 적용)
- retrofit을 생성자에서 전달받도록 변경
- 불필요한 serialize name 제거
- NumberFormatter 추가
Holder가 아닌, Presenter에서 데이터 가공하도록 변경
- RecyclerView 관련하여 xml에서 layout manager 설정하도록 변경
- 패딩 설정 시, 4단위로 적용하도록 변경

BeokBeok committed:
HW-3 코드 일괄 정리 (3차)

- BaseActivity 추가
- BaseFragment 에서 inflate 하도록 변경
- 외부에서 Layout ID 및 Drawable ID 설정하도록 변경

BeokBeok committed:
HW-3 RxJava 적용 및 패키지 정리

NackUn commented on BeokBeok/app/src/main/java/com/architecturestudy/upbitmarket/recyclerview/UpbitAdapter.kt:
```diff
@@ -0,0 +1,53 @@
+        parent: ViewGroup,
+        viewType: Int
+    ): ViewHolder {
+        val view = LayoutInflater.from(parent.context)
```
이 부분을 뷰홀더에서 만들어주고 한 줄로 축약하는건 어떠세요?

BeokBeok replied:
적용 하였습니다.
7a0b2e0

NackUn reviewed (changes requested):
리뷰해주신거 다 고치신 것 같고 추가적으로 적용하셨네요. 멋있습니다!!
Rx 쪽은 몰라서 리뷰를 못했네요. 고생하셨습니다.

BeokBeok committed:
HW-3 View Holder 에서 inflate 하도록 변경

- Review URL
https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/63#discussion_r297198729

NackUn commented on BeokBeok/app/src/main/java/com/architecturestudy/upbitmarket/recyclerview/UpbitAdapter.kt:
```diff
@@ -17,9 +16,7 @@ class UpbitAdapter(
         parent: ViewGroup,
         viewType: Int
     ): ViewHolder {
```
= ViewHolder(parent) 도 가능하지 않을까요?

BeokBeok replied:
구현부가 1 Line 인 경우에 대하여 적용하였습니다.
b7434f3

BeokBeok committed:
HW-3 구현부가 One Line 인 경우, 중괄호 제거

NackUn reviewed (dismissed):
고생하셨습니다!

sjjeong commented on BeokBeok/app/src/main/java/com/architecturestudy/upbitmarket/UpbitFragment.kt:
```diff
@@ -1,92 +1,93 @@
+        tv_market_krw.setOnClickListener { setViewColor(tv_market_krw) }
+        tv_market_btc.setOnClickListener { setViewColor(tv_market_btc) }
+        tv_market_eth.setOnClickListener { setViewColor(tv_market_eth) }
+        tv_market_usdt.setOnClickListener { setViewColor(tv_market_usdt) }
```
```kotlin
tv_market_usdt.setOnClickListener(this::setViewColor)
```
이렇게 사용하시면 더 좋습니다.

BeokBeok replied:
적용 하였습니다.
b2b3ae9

sjjeong commented on BeokBeok/app/src/main/java/com/architecturestudy/upbitmarket/UpbitFragment.kt:
```diff
@@ -1,92 +1,93 @@
         savedInstanceState: Bundle?
     ): View? {
-        val view = inflater.inflate(R.layout.fragment_upbit, container, false)
         presenter.start()
```
presenter.start()는 onViewCreated로 이동하는게 좋을 것 같습니다.

BeokBeok replied:
적용 하였습니다.
28a7f34

sjjeong commented on BeokBeok/app/src/main/java/com/architecturestudy/upbitmarket/recyclerview/UpbitAdapter.kt:
```diff
@@ -1,65 +1,51 @@
+    @LayoutRes
+    private val resource: Int
+) : RecyclerView.Adapter<UpbitAdapter.ViewHolder>() {
+    var marketPrice: List<Map<String, String>>? = null
```
```kotlin
val marketPrice = mutableListOf<Map<String, String>>()
```
으로 해주세요

BeokBeok replied:
적용 하였습니다.
6147124

sjjeong commented on BeokBeok/app/src/main/java/com/architecturestudy/upbitmarket/recyclerview/UpbitAdapter.kt:
```diff
@@ -1,65 +1,51 @@
-        private val netChange = itemView.findViewById<TextView>(R.id.tv_net_change)
-        private val tradingVal = itemView.findViewById<TextView>(R.id.tv_trading_value)
+    fun setMarketPrices(marketPrice: List<Map<String, String>>) {
+        this.marketPrice = marketPrice
```
외부와 객체를 공유하면 좋지 않습니다.
clear와 addAll을 사용해 Collection안에 있는 데이터만 넣어주도록 해주세요

BeokBeok replied:
적용 하였습니다.
6147124

sjjeong reviewed (changes requested):
고생하셨습니다!

BeokBeok committed:
HW-3 함수 참조 사용

- https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/63#discussion_r297720242

BeokBeok committed:
HW-3 presenter.start 호출부 변경 (onCreateView -> onViewCreated)

- 리뷰 URL
https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/63#discussion_r297720928

BeokBeok committed:
HW-3 market list 를 mutable로 변경

- 리뷰 URL
https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/63#discussion_r297721458
https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/63#discussion_r297721718

BeokBeok committed:
HW-3 retrofit gson 라이브러리 업데이트

NackUn reviewed (approved):
Approve 합니다!

sjjeong reviewed (approved):
LGTM ✊ ✌️ 👋
